contributors/design-proposals/api-paging.md

```diff
@@ -0,0 +1,148 @@
+# Reduce the amount of client and server memory allocated during large API lists
+
+On large clusters, performing API queries that return all of the objects of a given resource type (GET /api/v1/pods, GET /api/v1/secrets) can lead to significant variations in peak memory use on the server and contribute substantially to long tail request latency.
+
+When loading very large sets of objects -- some clusters are now reaching 100k pods or equivalent numbers of supporting resources -- the system must:
+
+* Construct the full range description in etcd in memory and serialize it as protobuf in the client
+  * Some clusters have reported over 500MB being stored in a single object type
```

Comment:
@liggitt Correct. I am just pointing out that clients that do not send `limit` will still cause issues. We should come up with a way that helps resolve the memory issue at the server.

For example: We could truncate results returned to clients which do not set the `limit` if the request is too large. Or we could return an `HTTP status 413` for large requests forcing the client to update and to use `limit`.